File: leetcode/454_4Sum_II.py
from typing import List
from unittest import TestCase

class Solution:
    def fourSumCount(self, nums1: List[int], nums2: List[int], nums3: List[int], nums4: List[int]) -> int:
        cnt, sums = 0, {}
        for a in nums1:
            for b in nums2:
                sums[a + b] = sums.get(a + b, 0) + 1
        for a in nums3:
            for b in  nums4:
                cnt += sums.get(- a - b, 0)
        return cnt
        
class FourSumCountTest(TestCase):
    def test_1(self):
        self.assertEqual(Solution().fourSumCount([1,2], [-2,-1], [-1,2], [0,2]), 2)

    def test_2(self):
        self.assertEqual(Solution().fourSumCount([0], [0], [0], [0]), 1)

File: leetcode/454_4Sum_II.rs
use std::collections::HashMap;

struct Solution {}

impl Solution {
    pub fn four_sum_count(nums1: Vec<i32>, nums2: Vec<i32>, nums3: Vec<i32>, nums4: Vec<i32>) -> i32 {
        let mut dic = HashMap::new();
        let mut cnt = 0;
        for a in &nums1 {
            for b in &nums2 {
                let value = 1 + match dic.get(&(a + b)) {
                    Some(value) =>  *value,
                    None => 0,
                };
                dic.insert(a + b, value);
            }
        }
        for a in &nums3 {
            for b in &nums4 {
                if let Some(value) = dic.get(&(- a - b)) {
                    cnt += value;
                }
            }
        }
        cnt
    }
}

mod four_sum_count_test {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::four_sum_count(vec![1,2], vec![-2,-1], vec![-1,2], vec![0,2]), 2);
    }

    #[test]
    fn test_2() {
        assert_eq!(Solution::four_sum_count(vec![0], vec![0], vec![0], vec![0]), 1);
    }
}